futures-util/src/task/local_waker_ref.rs

```diff
@@ -0,0 +1,140 @@
+#![allow(cast_ptr_alignment)] // clippy is too strict here
+
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::ptr::NonNull;
+use std::sync::Arc;
+use std::task::{LocalWaker, Waker, Wake, UnsafeWake};
+
+/// A [`LocalWaker`](::std::task::LocalWaker) that is only valid for a given lifetime.
+///
+/// Note: this type implements [`Deref<Target = LocalWaker>`](::std::ops::Deref),
+/// so it can be used to get a `&LocalWaker`.
+#[derive(Debug)]
+pub struct LocalWakerRef<'a> {
+    local_waker: LocalWaker,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a> LocalWakerRef<'a> {
+    /// Create a new [`LocalWakerRef`] from a [`LocalWaker`].
+    ///
+    /// Note: this function is safe, but it is generally only used
+    /// from `unsafe` contexts that need to create a `LocalWaker`
+    /// that is guaranteed not to outlive a particular lifetime.
+    pub fn new(local_waker: LocalWaker) -> Self {
+        LocalWakerRef {
+            local_waker,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<'a> Deref for LocalWakerRef<'a> {
+    type Target = LocalWaker;
+
+    fn deref(&self) -> &LocalWaker {
+        &self.local_waker
+    }
+}
+
+// Pointers to this type below are really pointers to `Arc<W>`
+struct ReferencedArc<W> {
+    _marker: PhantomData<W>,
+}
+
+unsafe impl<W: Wake + 'static> UnsafeWake for ReferencedArc<W> {
+    #[inline]
+    unsafe fn clone_raw(&self) -> Waker {
+        let me = self as *const ReferencedArc<W> as *const Arc<W>;
+        Waker::from(Arc::clone(&*me))
```

Comment:
Why not method call syntax?

```rust
Arc::clone(&*me).into()
```

Reply:
No particular reason.

Reply:
Hehe. Method call syntax would be nice then 😁

Reply:
Done.